### PR TITLE
Geometry optimise charged species

### LINKED ISSUE #96: Minimising species repeatedly eventually causes crash (GeometryOptimisation module)
Repeatedly clicking the species medic button eventually causes a crash when forces / dE become too small?

==61223== Invalid write of size 8
==61223==    at 0x56831F: Cell::addAtom(Atom*) (in /home/tris/src/dissolve/build/dissolve-gui)
==61223==    by 0x5789D4: Configuration::updateCellLocation(Atom*) (in /home/tris/src/dissolve/build/dissolve-gui)
==61223==    by 0x578A1C: Configuration::updateCellContents() (in /home/tris/src/dissolve/build/dissolve-gui)
==61223==    by 0x6EE5EC: GeometryOptimisationModule::energyAtGradientPoint(ProcessPool&, Configuration*, PotentialMap const&, double) (in /home/tris/src/dissolve/build/dissolve-gui)
==61223==    by 0x6EEB13: GeometryOptimisationModule::lineMinimise(ProcessPool&, Configuration*, PotentialMap const&, double, double&) (in /home/tris/src/dissolve/build/dissolve-gui)

### LINKED ISSUE #195: Can't geometry minimise charged species
Because a temporary Module layer is used when geometry-minimising Species, checks for charge neutrality are made and so (legitimately) charged molecules cannot be optimised.

Instead of using a temporary layer and all the fuss that involves, it may be beneficial to add additional functions to the GeometryOptimisation module to allow Species to be treated directly.

### PR DESCRIPTION
**Pull request by trisyoungs**
This PR addresses an issue with the GUI which prevents species with non-zero net charge being subjected to geometry minimisation, even though this is a completely legitimate thing to do.

### Cause
The `GeometryOptimisationModule` contains the core routines to perform geometry minimisation w.r.t. energy, however the focus of the module is on Configurations rather than Species. This resulted in the need to utilise a temporary instance of Dissolve to create and apply a module layer containing a GeometryOptimisationModule to a Configuration containing a single Molecule instance of the target Species to geometry optimise. Messy, but working. A previous commit put in place measures to prevent full simulations running on Configurations with net non-zero charge, as this is usually indicative of a problem with Species setup, and so the optimisation of the charged Species molecule is prevented.

### Resolution
The `GeometryOptimisationModule` has been rewritten to allow it to target either Configurations or Species (with no checks for charges on the latter), and associated functions added to `EnergyKernel` and `ForceKernel`. Other modifications allow the module to be used directly to perform the optimisation, rather than having to instantiate a temporary Dissolve object.

### Additional
Duplicate code has been removed in several areas.
Closes #195.
Closes #96.
Addresses undocumented bug whereby atomic coordinates were mirrored after each minimisation cycle.

### CODE REVIEW

File: src/classes/forcekernel.cpp [hunk 1]
```diff
@@ -522,6 +577,28 @@ void ForceKernel::forces(const Atom *onlyThis, const SpeciesBond *b, const Atom
...
+
+    // Get distance and normalise vector ready for force calculation
+    const auto distance = vecji.magAndNormalise();
+
+    // Determine final forces
+    vecji *= b->force(distance);
+
+    // Calculate forces
+    auto index = b->i()->index();
+    fx_[index] -= vecji.x;
```
Comment (rprospero) [inline src/classes/forcekernel.cpp:593]: Again, this might be something for a separate commit, but it would seem that having a single `f_` of type `Vec3<Double>` would save a lot of effort compared to having the three `fx_`, `fy_`, and `fz_` variables of type `Double`.  We could then take advantage of the operator overloading of `Vec3` to simplify this code greatly.

File: src/modules/forces/functions.cpp [hunk 2]
```diff
@@ -119,47 +71,78 @@ void ForcesModule::interatomicForces(ProcessPool &procPool, Configuration *cfg,
...
- */
+// Calculate interatomic forces within the specified Species
+void ForcesModule::interAtomicForces(ProcessPool &procPool, Species *sp, const PotentialMap &potentialMap, Array<double> &fx,
+                                     Array<double> &fy, Array<double> &fz)
+{
+    double scale, r, magjisq;
+    const auto cutoffSq = potentialMap.range() * potentialMap.range();
+    Vec3<double> vecij;
+    SpeciesAtom *j;
+    for (auto indexI = 0; indexI < sp->nAtoms() - 1; ++indexI)
```
Comment (rprospero) [inline src/modules/forces/functions.cpp:112]: Another one to flag for for_each_loop

File: src/classes/potentialmap.cpp [hunk 3]
```diff
@@ -140,7 +140,19 @@ double PotentialMap::energy(const Atom *i, const Atom *j, double r) const
...
+// Return energy between SpeciesAtoms at distance specified
+double PotentialMap::energy(const SpeciesAtom *i, const SpeciesAtom *j, double r) const
+{
+    // Check to see whether Coulomb terms should be calculated from atomic charges, rather than them being included in the
+    // interpolated potential
+    PairPotential *pp = potentialMatrix_.constAt(i->atomType()->index(), j->atomType()->index());
+    if (pp->includeCoulomb())
+        return pp->energy(r);
+    else
+        return (pp->energy(r) + pp->analyticCoulombEnergy(i->charge() * j->charge(), r));
```
[the remark this answers is not in the document]
Comment (trisyoungs) [reply src/classes/potentialmap.cpp:152]: Ternary operator good for me. Have changed accordingly.

File: src/modules/energy/functions.cpp [hunk 4]
```diff
@@ -21,19 +21,120 @@
...
+    Vec3<double> rI;
+    double scale, energy = 0.0;
+    const auto cutoff = potentialMap.range();
+
+    // Get start/end for loop
+    auto loopStart = procPool.twoBodyLoopStart(sp->nAtoms());
+    auto loopEnd = procPool.twoBodyLoopEnd(sp->nAtoms());
+
+    // Double loop over species atoms
+    for (auto indexI = loopStart; indexI <= loopEnd; ++indexI)
```
[the remark this answers is not in the document]
Comment (trisyoungs) [reply src/modules/energy/functions.cpp:73]: `NOTE` added.

File: src/modules/energy/functions.cpp [hunk 5]
```diff
@@ -104,43 +205,36 @@ double EnergyModule::intraMolecularEnergy(ProcessPool &procPool, Configuration *
...
-     *
-     * This is a parallel routine, with processes operating as process groups.
-     */
+    auto energy = 0.0;
 
-    // Create an EnergyKernel
-    EnergyKernel kernel(procPool, cfg, potentialMap);
+    // Loop over bonds
+    DynamicArrayConstIterator<SpeciesBond> bondIterator(sp->constBonds());
+    while (const SpeciesBond *b = bondIterator.iterate())
```
[the remark this answers is not in the document]
Comment (trisyoungs) [reply src/modules/energy/functions.cpp:215]: To be done as part of #313.

File: src/modules/geomopt/functions.cpp [hunk 6]
```diff
@@ -83,159 +103,59 @@ void GeometryOptimisationModule::sortBoundsAndEnergies(Vec3<double> &bounds, Vec
...
+    // Initialise working arrays for coordinates and forces
+    xRef_.initialise(sp->nAtoms());
+    yRef_.initialise(sp->nAtoms());
+    zRef_.initialise(sp->nAtoms());
+    xTemp_.initialise(sp->nAtoms());
+    yTemp_.initialise(sp->nAtoms());
+    zTemp_.initialise(sp->nAtoms());
+    xForce_.initialise(sp->nAtoms());
+    yForce_.initialise(sp->nAtoms());
+    zForce_.initialise(sp->nAtoms());
```
[the remark this answers is not in the document]
Comment (trisyoungs) [reply src/modules/geomopt/functions.cpp:156]: Had made a node to do that, but not in this commit as it will have fallout in a couple of other places.

File: src/classes/forcekernel.cpp (hunk 1 above)
Comment (trisyoungs) [reply src/classes/forcekernel.cpp:593]: Fair point.  Have to revisit the code to figure out why I thought individual arrays was a better solution.

File: src/modules/forces/functions.cpp (hunk 2 above)
Comment (trisyoungs) [reply src/modules/forces/functions.cpp:113]: `NOTE` added.
